modules/src/mock_context.rs

```diff
@@ -0,0 +1,199 @@
+use crate::ics02_client::client_def::{AnyClientState, AnyConsensusState};
+use crate::ics03_connection::context::{ConnectionKeeper, ConnectionReader};
+use crate::ics03_connection::error::Error as ICS3Error;
+use crate::ics23_commitment::commitment::CommitmentPrefix;
+use crate::ics24_host::identifier::{ClientId, ConnectionId};
+use std::collections::HashMap;
+use tendermint::block::Height;
+
+use crate::ics02_client::client_type::ClientType;
+use crate::ics02_client::context::ClientReader;
+use crate::ics03_connection::connection::ConnectionEnd;
+use crate::mock_client::header::MockHeader;
+use crate::mock_client::state::{MockClientRecord, MockClientState, MockConsensusState};
+use std::cmp::min;
+use std::convert::TryInto;
+
+/// Mock for a context. Used in testing handlers of all modules.
+#[derive(Clone, Debug)]
+pub struct MockContext {
+    /// Maximum size of the history.
+    max_history_size: usize,
+
+    /// Highest height of the headers in the history.
+    latest_height: Height,
+
+    /// A list of `max_history_size` headers, ascending order by their height (latest is last).
+    history: Vec<MockHeader>,
+
+    // All the connections in the store.
+    connections: HashMap<ConnectionId, ConnectionEnd>,
+
+    /// The set of all clients, indexed by their id.
+    clients: HashMap<ClientId, MockClientRecord>,
+
+    // Association between client ids and connection ids.
+    client_connections: HashMap<ClientId, ConnectionId>,
+}
+
+impl MockContext {
+    pub fn new(max_history_size: usize, latest_height: Height) -> Self {
+        // Compute the number of headers to store. If h is 0, nothing is stored.
+        let n = min(max_history_size as u64, latest_height.value());
+        MockContext {
+            max_history_size,
+            latest_height,
+            history: (0..n)
+                .rev()
+                .map(|i| MockHeader((latest_height.value() - i).try_into().unwrap()))
+                .collect(),
+            connections: Default::default(),
+            clients: Default::default(),
+            client_connections: Default::default(),
+        }
+    }
+
+    /// Internal interface for associating a connection to this context.
+    pub fn with_connection(
+        self,
+        connection_id: ConnectionId,
+        connection_end: ConnectionEnd,
+    ) -> Self {
+        let mut connections = self.connections.clone();
+        connections.insert(connection_id, connection_end);
+        Self {
+            connections,
+            ..self
+        }
+    }
+
```

Comment:
Wondering if we should check for sanity for the context, even if it's mock, just to make sure the tests don't initialize with a bad context. For example check that the client of a connection is present. Could also add the connection to client here.
I am also wondering if we should allow for a vector of connections, i.e. call it `with_connections()`.

Reply:
> I am also wondering if we should allow for a vector of connections, i.e. call it with_connections().

I'm slowly expanding the mocks to support more and more capabilities. So it's quite likely that we'll have such a function `with_connections`, which will enable more complex test cases. I can do that in the next iteration of mock  refurbishing (i.e., for #296).

> Wondering if we should check for sanity for the context, even if it's mock, just to make sure the tests don't initialize with a bad context. For example check that the client of a connection is present. Could also add the connection to client here.

Hmm, not sure I understand. Maybe this would be a good use case, indeed, but as I understand the mock context, its purpose is to support testing of handlers (and not more than that at the moment). So I would not add too much logic, validation, or any kind of support, unless it is to exercise more paths in the codebase.

Reply:
> Hmm, not sure I understand. Maybe this would be a good use case, indeed, but as I understand the mock context, its purpose is to support testing of handlers (and not more than that at the moment). So I would not add too much logic, validation, or any kind of support, unless it is to exercise more paths in the codebase.

Exactly, but in tests we always want to start with a correct context. This becomes harder to setup correctly with a more complex initial state.

Reply:
> I'm slowly expanding the mocks to support more and more capabilities. So it's quite likely that we'll have such a function with_connections, ....

This is would be the same as `with_connection`, just pass a slice and do a loop.

Reply:
Now I understand! Will track these two suggestions in the other iteration on this work (https://github.com/informalsystems/ibc-rs/issues/297#issue-717562826).